### DISCUSSION TITLE
Fleet Agents in Unhealthy state -- Elastic Defend no policy response.

### DISCUSSION THREAD

**Discussion by jaypfe** [Security-Onion-Solutions/securityonion#11148] [2.4] [answered]
### Version

2.4.10

### Installation Method

Security Onion ISO image

### Description

installation

### Installation Type

Distributed

### Location

on-prem with Internet access

### Hardware Specs

Exceeds minimum requirements

### Network Traffic Collection

other (please provide detail below)

### Network Traffic Speeds

Less than 1Gbps

### Status

Yes, all services on all nodes are running OK

### Salt Status

No, there are no failures

### Logs

No, there are no additional clues

### Detail

I currently have a manager and a search node deployed, no forward nodes or anything of the sort. Will be adding a fleet node soon. 

This is a new issue for me in GA. RC2 did not seem to have this issue. Upon installing elastic agent to an endpoint it will be healthy for about 5 seconds then so an unhealthy state. Looking into it, the elastic-defend-endpoints integration shows no policy response. I don't see really any logs coming off the machine either, just elalstic agent saying
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/a0819d1b-579e-4115-88ea-833bcb7e61a3)

Then looking the error logs in fleet I see 
09:01:47.153
elastic_agent
[elastic_agent][error] Component state changed system/metrics-default (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '24000': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Unit state changed beat/metrics-monitoring-metrics-monitoring-beats (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '13944': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Unit state changed system/metrics-default (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '24000': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Component state changed http/metrics-monitoring (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '18320': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Unit state changed http/metrics-monitoring-metrics-monitoring-agent (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '18320': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Unit state changed http/metrics-monitoring (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '18320': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Component state changed beat/metrics-monitoring (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '13944': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Unit state changed beat/metrics-monitoring (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '13944': failed to write connection information: write |1: The pipe has been ended.
09:01:47.153
elastic_agent
[elastic_agent][error] Unit state changed system/metrics-default-unique-system-metrics-input (STARTING->FAILED): Failed: failed to provide connection information to spawned pid '24000': failed to write connection information: write |1: The pipe has been ended.
09:01:59.879
elastic_agent
[elastic_agent][error] Spawned new component ENDPOINT_INTEGRATION_CONFIG-default: input not supported
09:01:59.879
elastic_agent
[elastic_agent][error] Spawned new unit ENDPOINT_INTEGRATION_CONFIG-default: input not supported
09:01:59.879
elastic_agent
[elastic_agent][error] Spawned new unit ENDPOINT_INTEGRATION_CONFIG-default-1cecd7f0-4071-11ee-9cf2-c12e9893e573: input not supported

Attempting to open the integration settings yields in this: 
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/ff4f4662-b03c-4833-87f6-5507014dd15e)

Some of this seems to be similar to #11084 but I am not on a mac.  This is on a win 11 endpoint and a server running server 2016

### Guidelines

- [X] I have read the above statement and can confirm my post is relevant to Security Onion 2.4.

**Comment by senatesan** [comment]
I uninstalled the Defend integration then added the transform role to my search nodes as explained in [#10349](https://github.com/Security-Onion-Solutions/securityonion/discussions/10349) and then reinstalled the Defend integration and it fixed my issue.

**Reply by jaypfe** [reply]
Are you making the change on the search nodes, or on the manager?

**Reply by senatesan** [reply]
I edited the `config.map.jinja` file on the manager

**Reply by senatesan** [reply]
To be more clear, I copied `/opt/so/saltstack/default/salt/elasticsearch/config.map.jinja` to `/opt/so/saltstack/local/salt/elasticsearch/config.map.jinja` on the manager and edited it to add the transform role to my search nodes

**Reply by jaypfe** [reply]
It appears I did something wrong as I can't find the defend integration in elastic anymore haha! 
I copied the file over, and I changed this line:  
 {% do ESCONFIG.elasticsearch.config.node.update({'roles': ['data', 'ingest']}) %}
to
 {% do ESCONFIG.elasticsearch.config.node.update({'roles': ['data', 'ingest', 'transform']}) %}
I am guessing this was the wrong edit.

**Reply by Odiesinfosec** [reply]
Have you already tried just making a new agent policy? Make it minimalist as possible and work back up. Fixed this issue for me

**Reply by jaypfe** [reply]
I tried to add the defend integration back to the endpoint-initial and it failed because it needed a transform role. 

I will try adding a new policy all together and see how that rides.

**Reply by senatesan** [reply]
This looks like it could be the right edit to me but I'm running a distributed mode. I am guessing you don't have search nodes?

**Reply by jaypfe** [reply]
No I do, that was the file I pulled from the master branch. If I pulled from the main branch it has the search node setting. 

I have lost the integration though, it no longer appears when I search for it in Kibana. Not sure how to go about fixing that other than nuking this deployment and starting again.

**Reply by senatesan** [reply]
In Kibana do you see Elastic Defend listed under Management > Integrations > Browse Integrations ?

**Reply by jaypfe** [reply]
lol nope 
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/107776997/e798a642-57dd-4ecd-b584-aaf748686cd0)

**Comment by Odiesinfosec** [comment]
I am getting very similar errors as well.. Unhealthy, No policy response. In fact, I am unable to update any of my agents to v 8.8.2, they just say 'updating' then will say 8.8.2, I refresh the page and its back to the old version... 
What is strange is that on my logs, all the "unhealthy" agents in fleet say **STARTING->HEALTHY** and I can't see any errors.
Running elastic-agent status shows all Healthy yet every single one of my agents, even when I freshly unenroll and reinstall immediately show "unhealthy" despite the agent reporting it is HEALTHY!

**Reply by Odiesinfosec** [reply]
Ok, I was able to fix it by creating a new minimalist policy with basic integrations. 
Elastic Defend (basic)
System (ultra basic)
windows (no perflogs)
Osquery 

I believe this has to do with messing with Elastic Defend integration, but not sure.

**Comment by defensivedepth** [comment, marked as answer]
@jaypfe @Odiesinfosec @senatesan We can confirm there is an issue with the endpoints-initial policy after the hotfix is applied - the Defend integration policy gets corrupted. You can fix this by doing the following:

```
Remove the Elastic Defend integration by navigating to Elastic Fleet -> Agent Policies -> endpoints-initial -> Actions -> Delete integration. Then run the following on the Manager:

sudo rm -f /opt/so/state/eaintegrations.txt
sudo so-elastic-fleet-integration-policy-load

If you are running a distributed grid with a Manager or ManagerSearch, you may also need to add the following role to Elasticsearch:
'transform'
sudo vi +24 /opt/so/saltstack/default/salt/elasticsearch/config.map.jinja 
so-elasticsearch-restart
```


If you are still having issues with it after this, please start a new Discussion and we can troubleshoot from there.

**Reply by senatesan** [reply]
@defensivedepth I'm just looking for clarification on this fix. I was adding the 'transform' role to my search nodes in my distributed environment. Does this role belong to the manager node instead?

**Reply by senatesan** [reply]
Update: I answered my own question. I saw in the Elastic's documentation it says:

```It is strongly recommended that dedicated transform nodes also have the remote_cluster_client role; otherwise, cross-cluster search fails when used in transforms.``` 

So it would be done on the manager as it has a `remote_cluster_client` role and the search nodes do not.

Ref: https://www.elastic.co/guide/en/elasticsearch/reference/8.8/transform-settings.html

**Reply by sjnhawk** [reply]
I'm having the same issue. Tried the fix above, but still getting unhealthy agents. Going into the Elastic Defend Integration, I now see this error:

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/67609108/b11857d0-15f5-4e23-b02a-ad425cefa529)

**Reply by defensivedepth** [reply]
@sjnhawk Please create a new Discussion and we can troubleshoot from there.
